diff -up util-linux-2.23.2/libblkid/src/partitions/gpt.c.kzak util-linux-2.23.2/libblkid/src/partitions/gpt.c
--- util-linux-2.23.2/libblkid/src/partitions/gpt.c.kzak	2014-09-25 10:36:26.761377688 +0200
+++ util-linux-2.23.2/libblkid/src/partitions/gpt.c	2014-09-25 10:36:56.912665364 +0200
@@ -332,7 +332,7 @@ static int probe_gpt_pt(blkid_probe pr,
 
 	blkid_probe_use_wiper(pr, lba * blkid_probe_get_size(pr), 8);
 
-	if (blkid_probe_set_magic(pr, lba << 9,
+	if (blkid_probe_set_magic(pr, blkid_probe_get_sectorsize(pr) * lba,
 			      sizeof(GPT_HEADER_SIGNATURE_STR) - 1,
 			      (unsigned char *) GPT_HEADER_SIGNATURE_STR))
 		goto err;
@@ -345,7 +345,8 @@ static int probe_gpt_pt(blkid_probe pr,
 	if (!ls)
 		goto err;
 
-	tab = blkid_partlist_new_parttable(ls, "gpt", lba << 9);
+	tab = blkid_partlist_new_parttable(ls, "gpt",
+				blkid_probe_get_sectorsize(pr) * lba);
 	if (!tab)
 		goto err;
 
